GeneratingDataset.py

```diff
@@ -2621,13 +2631,33 @@ def _collect_data(self):
     :return: entries
     :rtype: list[dict[str]]
     """
-    data = eval(self._read("%s.txt" % self._name))  # type: typing.List[typing.Dict[str]]
-    assert data and isinstance(data, list)
-    first_entry = data[0]
+    data_input = eval(self._read("%s.txt" % self._name))  # type: typing.List[typing.Dict[str]]
+    assert data_input and isinstance(data_input, list)
+    first_entry = data_input[0]
     assert isinstance(first_entry, dict)
     assert isinstance(first_entry["text"], str)
     assert isinstance(first_entry["duration"], float)
-    assert isinstance(first_entry["file"], str)
+
+    if "seq_name" in first_entry.keys():
+      assert isinstance(first_entry["seq_name"], str)
+    else:
+      assert isinstance(first_entry["file"], str)
+
+    segments = None
+    if self.segment_file:
+      segments = [s.strip() for s in open(self.segment_file)]
+
+    data = []
+    for entry in data_input:
+      if "seq_name" not in entry:
+        # use the filename as seq_name if it is not provided
+        entry['seq_name'] = entry['file']
+      if segments:
+        if entry['seq_name'] in segments:
+          data.append(entry)
+      else:
+        data.append(entry)
```

Comment:
Please avoid doing that for the default case (i.e. looping over all seqs). If the number of seqs is large (e.g. millions of seqs), this would take too much time, and is not quite necessary.

Reply:
But I need to iterate to check against the segment file, and doing it at access time is not saving time in the end, because the call would be done in each epoch, but the iteration at the beginning is done only once. The bigger problem is that segments is a list, it has to be changed to "set()" for major speed improvements. I created 5 million random strings, and iterated over 10 million random strings to check if being in the 5 million set, which took less than 10 seconds including converting from list to set. So time should not be an issue when using corpora less than 40k hours. (if e.g. 250.000 entries are 1000 hours)

Reply:
I'm not talking about that case. If you provide a segment file, yes, you can do the filtering here. However, otherwise, I would not iterate through it.